lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -1127,6 +1127,58 @@ struct RzToPhasedRx : public OpRewritePattern<quake::RzOp> {
   }
 };
 
+//===----------------------------------------------------------------------===//
+// U3Op decompositions
+//===----------------------------------------------------------------------===//
+
+// quake.u3(θ,ϕ,λ) target
+// ──────────────────────────────────
+// quake.rz(λ) target
+// quake.rx(π/2) target
+// quake.rz(θ) target
+// quake.rx(-π/2) target
+// quake.rz(ϕ) target
+struct U3ToRotations : public OpRewritePattern<quake::U3Op> {
+  using OpRewritePattern<quake::U3Op>::OpRewritePattern;
+
+  void initialize() { setDebugName("U3ToRotations"); }
+
+  LogicalResult matchAndRewrite(quake::U3Op op,
+                                PatternRewriter &rewriter) const override {
+    if (!op.getControls().empty())
+      return failure();
```

Comment:
Would the control case be the above transformation with the controls applied to each of the 5 gates?

Reply:
(I think the controls themselves are meant to be handled in the multicontrol-decompostion pass, so as long as this happens first, that pass may handle the rest of the transformation.